Live Nginx configuration update without reloading

ElvinEfendi opened the pull request:
Requires https://github.com/kubernetes/ingress-nginx/pull/2172 to be merged first and published the new Nginx image.

**What this PR does / why we need it**:
The PR implements live configuration update for Nginx upstreams/k8s endpoints. That means every time you deploy your app Nginx won't be reloaded. This is very important when you are front-ending many applications behind a single deployment of ingress-nginx and those app gets deployed frequently and have significant traffic.

I've mainly focused on laying out proper foundation for further development of live Nginx (re)configuration feature in ingress-nginx. The PR implements only Round Robin algorithm but can easily be extended to support more. I have another PR at https://github.com/kubernetes/ingress-nginx/pull/2167 to configure load balancing algorithm per ingress, that configuration will be used by Lua code as well to decide which LB algorithm to use for a given app(namespace/service).

The feature can be enabled by using `--enable-dynamic-configuration` command line flag. If you wanna see more Lua logs you can also set `error-log-level` to `info` to see what it does on endpoints changes.

In order to avoid JSON decoding in the request path, for every Nginx worker I spin up a periodic function that reads raw configuration from a shared Lua dictionary and decodes it into local cache per worker. Then balancer uses its local cache to get the list of endpoints and other backend configuration for given namespace/service.

**If you want to test** you can use `index.docker.io/elvinefendi/nginx-ingress-controller:0.0.1`

**Next steps**: Setup a testing framework for Lua code, add support for EWMA and more LB algorithms, add support for live update of secrets by using `certificate_by_lua`

**Which issue this PR fixes**: n/a

**Special notes for your reviewer**:
I'm aware that there's another similar PR being worked on at https://github.com/kubernetes/ingress-nginx/pull/2152. I was told about that only after I started working on this feature at https://github.com/Shopify/ingress/pull/16#issuecomment-370096531. There are significant differences in the implementation between the two PRs and I thought there could be value in creating this PR as well.

ElvinEfendi commented:
/assign @aledbf

aledbf commented:
@ElvinEfendi @valeriano-manassero can we get consensus about which version we are going to push?
We cannot have two version of the same feature.

ElvinEfendi commented:
@aledbf I’d love to hear what @valeriano-manassero thinks about going forward with this PR and collaboratively extending it later. I am going to work on this feature in coming weeks consistently and if you want I can put together a small RFC as well to discuss the future of this work in a more organized way.

I have already commented on other PR and raised my concerns. In my humble opinion the controller change in this PR is more concise and cleaner and the Lua implementation is more performant(no ‘json’ decoding in request path), correct(I use lua-resty-lock to avoid possible race condition between different Nginx workers accessing load balancing state in shared dictionary) and easier to extend(configuration.lua has its clear responsibility of accepting raw config piece per component and storing it in general shared dictionary which will be consumed by the respective middleware as needed. For example as of now ‘balancer.lua’ needs ‘backends’ only and it is responsible what needs to be done with that data alone. In future we can POST ‘servers’ config as well that will be consumed in a similar way by i.e ‘certificate.lua’ that serves certs dynamically etc).

We are also already running this code in our test cluster without any issue and also starting to migrate some less critical services to it(I will update with more data later).

Given the above compelling reasons I suggest we move forward with this PR.

Please let me know what do you think.

valeriano-manassero commented:
@aledbf @ElvinEfendi Sorry for the delay, I had a couple of very busy days.
I agree with you this PR is cleaner so it's ok for me to go on.
I will rebase further implementations I need on this PR when you'll merge in master.
Maybe a it's good if you want to discuss with me what I would like to implement for the future.
Thanks.

ElvinEfendi commented:
The CI should be fixed once https://github.com/kubernetes/ingress-nginx/pull/2172 gets merged and new Nginx image get published.

ElvinEfendi commented:
Thanks @valeriano-manassero! I would be happy to discuss that with you.

ElvinEfendi commented:
@aledbf could you push the new version of Nginx updated a thttps://github.com/kubernetes/ingress-nginx/pull/2172? Once I have that I'll fix the CI here.

aledbf commented:
@ElvinEfendi today I will publish the new nginx image

aledbf commented:
@ElvinEfendi please squash the commits (no more than three or four)

ElvinEfendi commented:
@aledbf looks like you've published the new Nginx image already, thanks!

CI is 🍏 now. Please let me know if you think this PR requires some more changes.

ElvinEfendi commented:
@aledbf is there anything blocking this PR from merging?

aledbf commented:
@ElvinEfendi only testing. If I don't find any issues I will merge this in two days.
Apologies but I've been really busy with other stuff.

ElvinEfendi commented:
@aledbf thanks for the update!

ElvinEfendi commented on internal/ingress/controller/template/template.go:
```diff
@@ -323,23 +323,28 @@ func buildProxyPass(host string, b interface{}, loc interface{}) string {
-			if isSticky(host, location, backend.SessionAffinity.CookieSessionAffinity.Locations) {
-				upstreamName = fmt.Sprintf("sticky-%v", upstreamName)
-			}
+	if !dynamicConfigurationEnabled {
```
This is currently buggy, it will ignore `https` proto when dynamic configuration is enabled. I have fixed it at https://github.com/Shopify/ingress/pull/29/files. @aledbf let me know if you want me include the fix in this PR or in a separate PR after this gets merged.

aledbf replied:
Please include the fix in this PR

aledbf commented on internal/ingress/controller/nginx.go:
```diff
@@ -745,3 +750,54 @@ func (n *NGINXController) setupSSLProxy() {
+		newBackends = append(newBackends, pcfg.Backends[i].DeepCopy())
+	}
+
+	n.runningConfig.Backends = []*ingress.Backend{}
```
Please do not change the state of `n.runningConfig`. Please make a copy and use that to make the comparison.

ElvinEfendi replied:
@aledbf notice that I'm changing only `n.runningConfig.Backends` and I have the copy of its original value at https://github.com/kubernetes/ingress-nginx/pull/2174/files#diff-cde3fffe2425ad7efaa8add1d05ae2c0R756 where I restore at https://github.com/kubernetes/ingress-nginx/pull/2174/files#diff-cde3fffe2425ad7efaa8add1d05ae2c0R772.

The idea is to make sure that the change between new and running config is only `Backends` change. Please let me know if you think there's a better way of doing this.

aledbf replied:
You should not change `n.runningConfig`. I prefer a full copy that we can be disposed of later.

ElvinEfendi replied:
Is there already a utility function to deep copy `runningConfig`?

ElvinEfendi commented:
@aledbf I included bug fix(https://github.com/kubernetes/ingress-nginx/pull/2174/commits/c7786543124ed1bc9f80d49e8c7f8a879302371c) and corresponding test cases(https://github.com/kubernetes/ingress-nginx/pull/2174/commits/3879d72212e25772f6b79e9a8f31db3129f7af6a).

I've also refactored(https://github.com/kubernetes/ingress-nginx/pull/2174/commits/2a9f13d3393568edd07fff1436c2dedee637f221) `IsDynamicallyConfigurable` implementation to not change `n.runningConfig` and improved the test(https://github.com/kubernetes/ingress-nginx/pull/2174/commits/534a3c6e0d915d3732f40fd7664a74128f21e396).

aledbf commented:
@ElvinEfendi please rebase and squash for the last review and merge

oilbeater commented on rootfs/etc/nginx/lua/configuration.lua:
```diff
@@ -0,0 +1,39 @@
+end
+
+function _M.call()
+  if ngx.var.request_method ~= "POST" or ngx.var.request_uri ~= "/configuration/backends" then
```
It's better to have a `GET` api to retrieve all configurations. As all configurations now store in memory not files, when something gets wrong, it will be difficult to debug if configuration is correct

ElvinEfendi commented:
Thanks @aledbf!

> If the dynamic configuration is enabled you need to skip the update of upstreams in the file nginx.conf (template)

I implemented it that way intentionally thinking that it would be more resilient to failures. But now that I think more about it is probably better to completely skip `OnUpdate` function call in `syncIngress` when dynamic reconfiguration is enabled. When we switch back to default behaviour the up-to-date configuration will be generated anyway.

> When you send the update to lua please include a log line like glog.V(2).Infof with the body you are sending. Without this is impossible to know what's the configuration in lua
Even better is you can add an additional route to get the content from lua.

👍 I'll do both.

aledbf commented:
> I implemented it that way intentionally thinking that it would be more resilient to failures. But now that I think more about it is probably better to completely skip OnUpdate function call in syncIngress when dynamic reconfiguration is enabled

Please keep in mind you only need to comment the upstream generation. We still need the reload of nginx when an Ingress is added/removed or a new annotation is configured

ElvinEfendi commented:
@aledbf addressed all your comments in the last 5 commits.

I have also pushed https://github.com/kubernetes/ingress-nginx/pull/2174/commits/66137baef94b889d91d667767c6577563a3bbdae to allow requests only from localhost to the Lua configuration endpoint.

aledbf commented:
@ElvinEfendi this is working just fine. Let's wait until tomorrow and I will merge this.

aledbf commented:
@ElvinEfendi after this PR we need to clean up the JSON is being sent to LUA because there's a lot of information we are not going to use.

aledbf commented:
/approve

aledbf commented:
@ElvinEfendi thanks!
